### PR TITLE
fix(run): resolve the bundled default checkpoint at runtime (BE-2994)

### PR DESCRIPTION
**Pull request by mattmillerai**
## ELI-5

`comfy run --prompt "a fox"` (no `--workflow`) runs a tiny built-in text2img
workflow. That workflow had one specific checkpoint filename baked in. If the
machine (or Comfy Cloud, or a fresh install) didn't have *that exact* file, the
run failed — often with a cryptic server-side error. This PR makes the CLI look
at what checkpoints the target actually has and use one, so a zero-config run
produces an image wherever *any* checkpoint exists — and gives a clear,
actionable error when there are truly none.

## What changed

1. **Fixed the stale pin.** `cql/data/default_text2img.json` node `"4"`
   `ckpt_name` is now `v1-5-pruned-emaonly-fp16.safetensors` (the file the
   gallery `default.json` and the text-to-image tutorial use), replacing the
   stale `v1-5-pruned-emaonly.ckpt`.

2. **Runtime checkpoint resolution.** New pure, offline-testable helper
   `resolve_default_checkpoint(workflow, object_info)` in
   `cql/default_workflow.py`:
   - pinned checkpoint present in the target's enum → **no change**;
   - pinned absent but the target has ≥1 checkpoint → **substitute the first
     available** and emit a note;
   - target positively enumerates **zero** checkpoints → flag it (hard error, below);
   - enum absent / `object_info` unfetchable (`{}`) → **fail open** (unchanged
     behavior — submit and let the server answer).

   Wired into **both** run paths (local `execute` + `execute_cloud`) after
   `object_info` is fetched and before `_preflight_validate`. It only acts on
   the bundled default graph (`workflow_name == "default_text2img"`) **and** is
   skipped entirely when the user pinned the checkpoint via `--set checkpoint=…`
   / `--set 4.ckpt_name=…` (honored verbatim; a `checkpoint_user_set` flag is
   threaded through the `preloaded` tuple).

3. **Actionable no-checkpoint error.** When the target positively has zero
   checkpoints, a new registered `no_checkpoint_available` error (exit 1) with a
   download hint (local) / gallery-template guidance (cloud) replaces the cryptic
   server reject. Fail-open behavior is preserved when `object_info` can't be
   fetched at all.

Substitution is surfaced as a `checkpoint_substituted` NDJSON event (`--json`)
and a one-line pretty notice otherwise.

## Acceptance criteria

- [x] `default_text2img.json` uses `v1-5-pruned-emaonly-fp16.safetensors`; fixtures/tests updated.
- [x] Missing-default-but-has-others → succeeds using the first available + emits a note; `--set checkpoint=<name>` always honored verbatim.
- [x] Default present → unchanged, no note.
- [x] Zero checkpoints (empty enum) → `no_checkpoint_available` + exit 1; `object_info` `{}` → fail-open submit (unchanged).
- [x] Both `--where local` and `--where cloud`.
- [x] New unit tests for `resolve_default_checkpoint` (present / substituted / empty-enum / user-pinned) + wiring tests; `ruff`/`pytest` green.

## Testing

`uv run --extra dev pytest tests/comfy_cli/cql tests/comfy_cli/command/test_run_prompt.py tests/comfy_cli/command/test_run.py tests/comfy_cli/command/test_run_json.py tests/comfy_cli/command/test_run_cli.py tests/comfy_cli/output` → **488 passed**. `ruff format --check` clean on all touched files.

## Judgment calls / notes

- **Return shape.** The ticket suggested `resolve_default_checkpoint(...) -> (workflow, note|None)`, but a bare `note` can't distinguish "pinned present (fine)" from "enum positively empty (error)" — both would be `None`. I returned a small `CheckpointResolution` dataclass (`note` / `substituted_to` / `no_checkpoint`) so the caller can drive step 3 precisely and only hard-error when it *positively* knows the enum is empty.
- **`--print-prompt` is unchanged.** The dry-run returns before the server round-trip, so it prints the bundled pin, not the runtime-resolved checkpoint (consistent with `--print-prompt`'s existing "no server hit" contract). The `prompt_preview` event (emitted before resolution) likewise shows the pre-substitution graph; the explicit `checkpoint_substituted` event documents the swap in the stream.
- **Event schema.** `checkpoint_substituted` is not added to `schemas/run_event.json`'s `type` enum — that enum is already non-exhaustive (`prompt_preview`/`converted` aren't in it either; `additionalProperties: true`), so I matched the established pattern rather than partially expanding it.
- **Not a capability-denial regression (falsification).** The new `no_checkpoint_available` deny path is capability-*enabling* on net: runtime resolution makes `--prompt` succeed wherever any checkpoint exists (the opposite of a dead-end). The error fires *only* in the provably-empty-enum case — a state where the workflow genuinely cannot run (no checkpoint to load) — and its hint redirects to the real fix (install a model / gallery template / `--set`). Fail-open is preserved for every ambiguous case.

## Out of scope

Restructuring the bundled default to the gallery's newer Z-Image-Turbo graph (multi-file diffusion) is a larger, separate change and a product call on the blessed default model. Runtime resolution deliberately sidesteps that choice.

### CODE REVIEW

File: comfy_cli/command/run/__init__.py [hunk 1]
```diff
@@ -593,6 +606,13 @@ def execute_cloud(
     except Exception:  # noqa: BLE001
         cloud_object_info = {}
 
+    # Runtime checkpoint resolution for the bundled `--prompt` default (mirrors
+    # the local path): swap the pinned checkpoint for one Comfy Cloud actually
+    # has, or hard-error if it enumerates none. Guarded to the bundled default
+    # and skipped when the user pinned the checkpoint explicitly.
+    if preloaded is not None and workflow_name == "default_text2img" and not checkpoint_user_set:
+        _resolve_default_checkpoint_or_exit(renderer, parsed_workflow, cloud_object_info, where="cloud")
```
[the remark this answers is not in the document]
Comment (mattmillerai) [reply comfy_cli/command/run/__init__.py]: Fixed in 93cfdfb — `_resolve_default_checkpoint_or_exit` now fails open on an empty enum for `where="cloud"` (returns without erroring), since Comfy Cloud provisions models per-job at runtime. The hard `no_checkpoint_available` error is retained only for the local path, where the enum reflects what's actually installed. Added `test_cloud_empty_enum_fails_open`.

File: comfy_cli/cql/default_workflow.py [hunk 2]
```diff
@@ -219,3 +225,112 @@ def build_default_workflow(*, prompt: str | None = None, overrides: list[str] |
...
+
+def _checkpoint_enum(object_info: dict) -> list | None:
+    """Return the ``CheckpointLoaderSimple.ckpt_name`` option list from
+    ``object_info``, or ``None`` when it isn't enumerated at all.
+
+    The raw object_info shape is ``ckpt_name: [[<name>, …], {opts}]`` — the
+    option list is element 0. ``None`` (not an empty list) means "can't tell"
+    so the caller can distinguish a positively-empty enum from an absent one.
+    """
+    node = object_info.get("CheckpointLoaderSimple")
```
[the remark this answers is not in the document]
Comment (mattmillerai) [reply comfy_cli/cql/default_workflow.py:304]: Fixed in 93cfdfb — `_checkpoint_enum` now returns `None` (fail open) when `object_info` isn't a dict, mirroring `Graph.from_object_info`'s guard, so a non-object `/object_info` payload no longer raises `AttributeError`. Added `test_non_dict_object_info_fails_open`.

File: comfy_cli/command/run/__init__.py [hunk 3]
```diff
@@ -214,6 +218,14 @@ def execute(
     # extra_data so the partner node finds it server-side — same shape
     # the cloud submit path uses.
     object_info = _fetch_object_info(host, port)
+
+    # Runtime checkpoint resolution for the bundled `--prompt` default: swap the
+    # pinned checkpoint for one the local server actually has (or hard-error if
+    # it has none). Guarded to the bundled default graph and skipped when the
+    # user pinned the checkpoint explicitly (honor it; let preflight reject it).
+    if preloaded is not None and workflow_name == "default_text2img" and not checkpoint_user_set:
+        _resolve_default_checkpoint_or_exit(renderer, workflow, object_info, where="local")
```
[the remark this answers is not in the document]
Comment (mattmillerai) [reply comfy_cli/command/run/__init__.py]: Fixed in 93cfdfb — checkpoint resolution now runs BEFORE the `prompt_preview` event in the submit flow (both local and cloud), so the streamed audit trail reflects the graph actually submitted. `--print-prompt` intentionally stays a documented no-server-hit dry-run (see its help text) and prints the graph as-is, before any server-dependent resolution.

File: comfy_cli/cql/default_workflow.py [hunk 4]
```diff
@@ -219,3 +225,112 @@ def build_default_workflow(*, prompt: str | None = None, overrides: list[str] |
...
+    enum = _checkpoint_enum(object_info)
+    if enum is None:
+        # Not enumerated (fresh/unfetched object_info) — fail open.
+        return workflow, CheckpointResolution()
+    if not enum:
+        # Positively empty: the target has zero checkpoints installed.
+        return workflow, CheckpointResolution(no_checkpoint=True)
+
+    pinned = inputs.get("ckpt_name")
+    if pinned in enum:
```
[the remark this answers is not in the document]
Comment (mattmillerai) [reply comfy_cli/cql/default_workflow.py:364]: Fixed in 93cfdfb — added a basename fallback before substituting: a pinned bare filename now matches an enum entry living in a subfolder (e.g. `SD1.5/v1-5-…safetensors`) and rewrites `ckpt_name` to that full path with no substitution. Added `test_pinned_matched_by_basename_in_subfolder`.

File: comfy_cli/command/run/preflight.py [hunk 5]
```diff
@@ -111,6 +111,57 @@ def _preflight_validate(renderer, workflow: dict, object_info: dict, *, target_l
...
+            hint=hint,
+            details={"where": where},
+        )
+        raise typer.Exit(code=1)
+
+    if res.note:
+        # Event fires in NDJSON/stream mode only; the pretty line covers humans.
+        renderer.event("checkpoint_substituted", message=res.note, checkpoint=res.substituted_to, where=where)
+        if renderer.is_pretty():
+            pprint(f"[yellow]⚠ {res.note}[/yellow]")
```
[the remark this answers is not in the document]
Comment (mattmillerai) [reply comfy_cli/command/run/preflight.py]: Fixed in 93cfdfb — the pretty substitution notice now wraps `res.note` with `rich.markup.escape` so a target-provided checkpoint name containing Rich tags can't inject terminal markup.

File: comfy_cli/command/run/preflight.py [hunk 6]
```diff
@@ -111,6 +111,57 @@ def _preflight_validate(renderer, workflow: dict, object_info: dict, *, target_l
...
+                f"the bundled default text2img workflow needs a checkpoint, but {target_label} has none installed"
+            ),
+            hint=hint,
+            details={"where": where},
+        )
+        raise typer.Exit(code=1)
+
+    if res.note:
+        # Event fires in NDJSON/stream mode only; the pretty line covers humans.
+        renderer.event("checkpoint_substituted", message=res.note, checkpoint=res.substituted_to, where=where)
```
[the remark this answers is not in the document]
Comment (mattmillerai) [reply comfy_cli/command/run/preflight.py:174]: Deferring this one as out-of-scope for this PR (leaving unresolved for visibility). `renderer.event()` is a no-op whenever `is_stream()` is false, so single-envelope JSON drops *every* event, not just `checkpoint_substituted` — surfacing events in the final envelope is a general renderer capability, not something to special-case for this one note. The substitution is fully surfaced today in NDJSON stream mode (`--json-stream`) and pretty mode. Threading per-event notes into the final JSON envelope is a broader renderer enhancement worth its own change rather than bolting a bespoke path onto the checkpoint resolver here.

File: comfy_cli/cql/default_workflow.py [hunk 7]
```diff
@@ -219,3 +225,112 @@ def build_default_workflow(*, prompt: str | None = None, overrides: list[str] |
...
+        return workflow, CheckpointResolution()
+
+    replacement = enum[0]
```
[the remark this answers is not in the document]
Comment (mattmillerai) [reply comfy_cli/cql/default_workflow.py:378]: Acknowledged as a real tradeoff, but declining the family-match heuristic here (leaving unresolved for visibility). Reliably preferring a same-architecture checkpoint needs metadata the `ckpt_name` enum doesn't expose — filenames alone don't distinguish SD1.5 / SDXL / Flux, so any name-based heuristic would be brittle and could itself select a wrong model. The substitution already emits a clear `checkpoint_substituted` note and points at `--set checkpoint=<name>`, and the new basename fallback (sibling thread) eliminates the most common false substitution. First-available-with-a-visible-note is the safer default; a metadata-driven family match belongs in a separate enhancement.

File: comfy_cli/command/run/preflight.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mattmillerai) [reply comfy_cli/command/run/preflight.py:174]: Recorded as a tracked follow-up rather than left as a loose thread, so resolving here.

To restate the deferral: this isn't specific to `checkpoint_substituted` — `renderer.event()` is a no-op whenever `is_stream()` is false, so single-envelope JSON drops **every** event, not just this one. The fix is a general renderer capability (accumulate events, attach them to the terminal envelope), not a special case bolted onto the checkpoint path — doing it locally here would add a second, inconsistent way to report events.

The follow-up captures the root cause, the reachability (`COMFY_OUTPUT=json` / agent auto-detect without the local `--json` that calls `force_stream()`), and the open schema question (new `events`/`warnings` key, interaction with `comfy_cli/schemas/*.json` + `test_envelope_schemas.py`, all events vs. a curated set). It is intentionally **not** labelled `agent-ok` — that schema decision wants a human first.

The substitution stays fully surfaced today in NDJSON/stream and pretty modes.

File: comfy_cli/cql/default_workflow.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mattmillerai) [reply comfy_cli/cql/default_workflow.py:378]: Recorded as a tracked follow-up rather than left as a loose thread, so resolving here.

The tradeoff is real and the ticket keeps it visible: the bundled graph's KSampler/latent settings are SD1.5-tuned, so an `enum[0]` that happens to be SDXL/Flux yields a poor result — and on paid cloud, bills compute for a model the user never picked.

Still declining a name-based family heuristic *in this PR*: the `ckpt_name` enum exposes only filenames, which don't reliably encode architecture, so such a heuristic would be brittle and could itself confidently pick the wrong model — worse than today's honest "first available + told you about it" with a `--set checkpoint=<name>` override. This PR's scope is removing the hard pin on a checkpoint that may not exist; picking the *right* substitute needs real metadata (safetensors header / model config) or a UX call (surface the enum on cloud instead of auto-substituting).

Follow-up lists those options and is intentionally **not** `agent-ok` — it needs a design decision, not a blind ship.

**Comment by mattmillerai** [comment]
Reviews resolver pass — all 7 review threads are now resolved, and no code changes were needed.

**Review threads:** 5 were already fixed in 93cfdfb. The 2 remaining were deferred/declined with reasoning; rather than leave them as loose unresolved threads, both are now recorded as tracked follow-ups (renderer events in single-envelope JSON; architecture-aware checkpoint substitution) and resolved. Neither is tagged for autonomous pickup — each needs a design decision first.

**Merge state:** `MERGEABLE`, no conflict with `main`.

**The failing `test` check is NOT from this PR — it is a repo-wide CI failure.**

`test-windows.yml` ("Windows Specific Commands") has failed **20 of 20 recent runs across every branch**, including many PRs sharing no code with this one (be-3271, be-3272, be-3265, be-3268, be-3266, be-3270, be-2999, be-3000, …). Last green run on `main` was 2026-03-11 (1405d30f). A `gh run rerun --failed` reproduced it identically (job 87814223657), so it is deterministic, not a flake.

Root cause: the workflow `pip install -e .` into `venv`, then runs `comfy install --fast-deps` **from that same venv**. `uv` tries to replace pydantic-core, but the live `comfy` process has already imported pydantic (via `mixpanel`), so Windows holds the loaded DLL open:

```
error: failed to remove file `...\venv\Lib\site-packages\pydantic_core/_pydantic_core.cp312-win_amd64.pyd`: Access is denied. (os error 5)
→ ImportError: cannot import name '__version__' from 'pydantic_core' (unknown location)
```

This PR touches only `cmdline.py`, `command/run/{__init__,preflight}.py`, `cql/default_workflow.py`, `cql/data/default_text2img.json`, `error_codes.py` + tests — nothing in the install/dependency path. Filed as a separate high-severity follow-up; fixing CI infra inside this behavioural PR would mask a repo-wide problem.

**Verified locally:** 63/63 PR tests pass (`test_run_prompt.py`, `test_default_workflow.py`), `ruff check` clean. Full suite: 2588 passed / 9 failed — the 9 (`test_config_parser.py`, `test_node_init.py`) are pre-existing, fail identically with `main`'s copies of those files, are untouched by this PR, and pass in CI on all platforms (local git-env artifact).

All other checks green: ruff, build, CodeQL, GPU runners, and Tests on macos/ubuntu/**windows** 3.10.

**Comment by mattmillerai** [comment]
🤖 The reviews loop filed Linear follow-up ticket(s) for review thread(s) deferred as out of scope for this PR:
- <https://linear.app/comfyorg/issue/BE-3281|BE-3281> — Fix repo-wide Windows CI failure: uv cannot replace pydantic_core .pyd locked by the running comfy process
- <https://linear.app/comfyorg/issue/BE-3282|BE-3282> — Surface renderer events (e.g. checkpoint_substituted) in single-envelope JSON output
- <https://linear.app/comfyorg/issue/BE-3283|BE-3283> — Prefer an architecture-compatible checkpoint instead of enum[0] when the bundled default is missing

**Comment by bigcat88** [comment]
This PR currently **conflicts with `main`** — GitHub reports `mergeable: CONFLICTING`, so it needs a rebase before I can review it and I'm skipping it in the current review sweep.

Please rebase (or merge `main` in) and I'll pick it up on the next pass. `main` moved a fair bit in the last day, including #614 (ANSI sanitisation across the pretty-print call sites) and #628 (the duplicate `server_died` error-code fix that had `main` red), so a refresh may also clear unrelated CI noise on this branch.

**Comment by mattmillerai** [comment]
Rebased on \`main\` — the merge conflict blocking review (in \`comfy_cli/cmdline.py\` and \`tests/comfy_cli/command/test_run_prompt.py\`) is resolved. Both conflicts were between this branch's runtime checkpoint-resolution work and \`main\`'s simpler static "needs a checkpoint" warning notice (\`TestDefaultCheckpointNotice\`); kept this branch's runtime resolution (which supersedes the static warning) and dropped main's now-superseded notice/tests accordingly.

Verified on the merged tree: full suite (3431 passed / 37 skipped), \`ruff check\`/\`ruff format --diff\` clean against the CI-pinned ruff 0.15.15.

The failing Windows `test` check is the pre-existing repo-wide CI issue already tracked as BE-3281 (uv can't replace a locked `.pyd` on Windows) — unrelated to this PR's changes, not re-fixed here per the prior resolver pass's reasoning (fixing CI infra inside this behavioral PR would mask the repo-wide problem). All 7 review threads were already resolved in a prior pass.

@bigcat88 conflict is cleared — ready for another look whenever convenient.

File: comfy_cli/command/run/__init__.py [hunk 8]
```diff
@@ -720,6 +738,23 @@ def execute_cloud(
...
 
+    # Cloud path uses cached/bundled object_info (no live server needed). Load
+    # it up front so checkpoint resolution can run BEFORE the preview/print
+    # below — the audit trail must advertise the graph we actually submit.
+    try:
+        from comfy_cli.cql.engine import _load_from_target
+
+        cloud_object_info = _load_from_target(mode="cloud")
+    except Exception:  # noqa: BLE001
+        cloud_object_info = {}
```
[the remark this answers is not in the document]
Comment (dynation7) [reply comfy_cli/command/run/__init__.py]: Love it

File: comfy_cli/cmdline.py [hunk 9]
```diff
@@ -946,37 +946,18 @@ def run(
...
-                # value containing [brackets] can't be read as rich markup.
-                rprint(
-                    f"[dim]Using the bundled default text2img workflow — it needs the[/dim] "
-                    f"[bold]{_escape(ckpt)}[/bold] [dim]checkpoint {where_ckpt}. "
-                    f"Override it with --set checkpoint=<name>.[/dim]"
-                )
+            # If the user pinned the checkpoint (--set checkpoint=… / 4.ckpt_name=…),
+            # honor it verbatim: runtime resolution is skipped downstream.
+            checkpoint_user_set = overrides_set_checkpoint(set_overrides, injected)
+            preloaded = (injected, "default_text2img", False, checkpoint_user_set)
```
[the remark this answers is not in the document]
Comment (mattmillerai) [reply comfy_cli/cmdline.py:971]: Fixed in ee688b8 — the help text was stale on both counts. Now names `v1-5-pruned-emaonly-fp16.safetensors` and describes the runtime substitution instead of the old "NOT downloaded for you" story.

File: comfy_cli/command/run/__init__.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mattmillerai) [reply comfy_cli/command/run/__init__.py]: Fixed in ee688b8, and this was worth more than trivial — `_load_from_target` is a live, uncached HTTPS fetch (bounded 64 MB read), so the UI-format cloud path really was paying for two round-trips.

Took your parenthetical rather than the `locals()` poke: `cloud_object_info` is declared before the `is_ui` branch and bound there when the conversion snapshot is loaded. One deviation — the sentinel is `None`, not falsiness. `if not cloud_object_info` would re-fetch when the first fetch legitimately returned an empty snapshot, which is exactly the case where the second fetch is guaranteed to be useless too.

File: comfy_cli/cql/default_workflow.py [hunk 10]
```diff
@@ -239,3 +245,128 @@ def build_default_workflow(*, prompt: str | None = None, overrides: list[str] |
...
+    # ComfyUI enumerates checkpoints by their path relative to the models dir,
+    # so a pinned bare filename won't exact-match the same file living in a
+    # subfolder (e.g. ``SD1.5/v1-5-…safetensors``). Prefer a basename match to
+    # the *intended* checkpoint before falling back to an arbitrary substitute.
+    if isinstance(pinned, str):
+        pinned_base = pinned.rsplit("/", 1)[-1]
+        for entry in enum:
+            if isinstance(entry, str) and entry.rsplit("/", 1)[-1] == pinned_base:
+                inputs["ckpt_name"] = entry
+                return workflow, CheckpointResolution()
```
[the remark this answers is not in the document]
Comment (mattmillerai) [reply comfy_cli/cql/default_workflow.py:376]: Fixed in ee688b8 — good catch, and the reasoning holds: `folder_paths` builds these names with `os.path.relpath`, so the separator belongs to the *server*, not to us. A Windows host enumerating `SD1.5\v1-5-pruned-emaonly-fp16.safetensors` would have missed the intended checkpoint and substituted `enum[0]` with a warning, which is the worst of both worlds.

Extracted the normalization into a module-level `_basename` helper (used on both sides) rather than a closure, and added `test_pinned_matched_by_basename_windows_separator` per your suggestion. Since it is the server's separator, the test is pure string logic and holds on any client OS.

File: tests/comfy_cli/cql/test_default_workflow.py [hunk 11]
```diff
@@ -182,3 +190,113 @@ def test_corrupt_bundle_surfaces_controlled_error(self, monkeypatch):
...
+        out, res = resolve_default_checkpoint(wf, oi)
+        assert out[CHECKPOINT_LOADER_ID]["inputs"]["ckpt_name"] == subfoldered
+        assert res.note is None
+        assert res.substituted_to is None
+        assert res.no_checkpoint is False
+
+    def test_object_info_without_checkpoint_node_fails_open(self):
+        # object_info present but no CheckpointLoaderSimple → can't tell → fail open.
+        wf = build_default_workflow(prompt="fox")
+        out, res = resolve_default_checkpoint(wf, {"KSampler": {"input": {"required": {}}}})
```
[the remark this answers is not in the document]
Comment (mattmillerai) [reply tests/comfy_cli/cql/test_default_workflow.py]: Fixed in ee688b8 — `_, res`, matching the sibling test.

File: tests/comfy_cli/command/test_run_prompt.py [hunk 12]
```diff
@@ -16,7 +16,16 @@
...
+from comfy_cli.cql.default_workflow import (
+    CHECKPOINT_LOADER_ID,
+    DEFAULT_CHECKPOINT_NAME,
+    POSITIVE_PROMPT_ID,
+    build_default_workflow,
+)
+
+
+def _object_info_with_checkpoints(names):
+    return {"CheckpointLoaderSimple": {"input": {"required": {"ckpt_name": [list(names), {}]}}}}
```
[the remark this answers is not in the document]
Comment (mattmillerai) [reply tests/comfy_cli/command/test_run_prompt.py:28]: Declining this one, deliberately.

The "duplication" is a single dict literal — one line of body — and the two copies live in different test packages (`tests/comfy_cli/command/` and `tests/comfy_cli/cql/`). Their only shared ancestor is `tests/comfy_cli/conftest.py`, so hoisting means adding a fixture to the namespace of *every* test under `tests/comfy_cli/` to serve two modules, and it puts the stub `object_info` shape a directory away from both tests that read it.

That trades a one-line duplication for a wider blast radius and worse locality. Self-contained stubs are the better default in test code: when one of these tests fails, the shape it asserts against is right there. Also worth noting the two stubs are not actually coupled — if the cloud path's enum shape ever diverges from the local one, sharing them would be the bug, not the fix.

Not filing a follow-up either: this is cosmetic with no behavioral risk, and your own rating (Trivial / Low value) agrees.

**Comment by mattmillerai** [comment]
Reviews resolver pass — CodeRabbit's 5 threads are all resolved; 4 fixed in ee688b8, 1 declined with reasoning.

**Fixed**
- **`--prompt` help text was stale on both counts** (`comfy_cli/cmdline.py`) — still named `v1-5-pruned-emaonly.ckpt` and still claimed the checkpoint "is NOT downloaded for you", with no mention of the runtime substitution this PR adds.
- **Cloud path fetched `/object_info` twice** (`comfy_cli/command/run/__init__.py`) — once to lower UI→API, then again for checkpoint resolution + preflight. `_load_from_target` is a live, uncached HTTPS fetch, so this was a genuine second round-trip, not just tidiness. Now loaded once and shared. Sentinel is `None` rather than falsiness, so a fetched-but-empty snapshot does not trigger a redundant refetch.
- **Basename match split on `/` only** (`comfy_cli/cql/default_workflow.py`) — `folder_paths` builds relative names with the *server's* separator, so a Windows host enumerating `SD1.5\v1-5-pruned-emaonly-fp16.safetensors` missed the intended checkpoint and substituted `enum[0]` with a warning. Normalized via a module-level `_basename` helper, with a regression test.
- Dropped an unused `out` binding in a resolver test (Pylint W0612).

**Declined:** hoisting the two-line `_object_info_with_checkpoints` stub into a shared conftest. The two copies live in different test packages whose only shared ancestor is `tests/comfy_cli/conftest.py`, so it would widen a fixture's blast radius to every test under `tests/comfy_cli/` to serve two modules, and move the asserted shape away from the tests that read it. Reasoning is on the thread.

**Verification:** full suite 3432 passed / 37 skipped; `ruff format` clean on all touched files. (Local `ruff check` flags UP038 in `default_workflow.py`, but that is pre-existing on untouched code and absent under the newer CI-pinned ruff — `ruff_check` is green.)

**Status:** all 15 CI checks pass, including `Run Tests on Multiple Platforms (windows-latest)` — the repo-wide Windows failure tracked as BE-3281 is no longer firing on this branch. `MERGEABLE` against `main`, no conflict. Not stacked. Awaiting human review — I do not merge.

**Comment by mattmillerai** [comment]
Reviews resolver pass — `main` had moved forward again since the last pass (adding `allow_spend` to `execute`/`execute_cloud`), reintroducing a merge conflict with this branch's runtime-checkpoint `preloaded` 4-tuple param. Merged `main` and kept both additions side by side (`preloaded: tuple[dict, str, bool, bool] | None` + `allow_spend: bool = False`) in both function signatures; `cmdline.py`'s call sites already threaded both correctly post auto-merge.

No unresolved review threads found (all 7 CodeRabbit threads from the prior pass were already resolved). Verified on the merged tree: full suite (3669 passed / 37 skipped), `ruff format --check` clean on touched files. Local `ruff check` flags 17 pre-existing UP038 issues, all confirmed present on untouched `origin/main` code — not introduced here.

`MERGEABLE` against `main`, not stacked. Awaiting human review — I do not merge.

**Review by skishore23 — APPROVED** [review]
Approving — this is the strongest PR in the stack so far. The fail-open discipline is consistent and the edge cases are genuinely thought through, not just asserted.

## Verified (merged `origin/main` into the head — clean merge)

- **Tests: 429 passed** across `tests/comfy_cli/cql`, `test_run_prompt.py`, `test_run.py`, `test_run_json.py`, `test_run_cli.py`. CI fully green.
- **Constant/data sync is enforced, not just documented.** `DEFAULT_CHECKPOINT_NAME` and `data/default_text2img.json` node `"4"` both read `v1-5-pruned-emaonly-fp16.safetensors`, and `test_default_workflow.py:74-75` pins *both* — so the "keep the two in sync" comment has a real guard behind it. The only surviving `v1-5-pruned-emaonly.ckpt` references are in `test_models.py`, where it's a genuine HuggingFace download URL, not the pin.
- **The `preloaded` 3-tuple → 4-tuple widening is complete.** Grepped every call site: one producer (`cmdline.py:978`), two consumers (`execute`, `execute_cloud`), all tests updated. No stale 3-tuple unpack left to blow up at runtime.
- **`--print-prompt` still makes no server hit.** The fetch moved earlier but sits behind `if not print_prompt:`, and the dry-run returns before `object_info` is read. Contract preserved.
- **Moving `_fetch_object_info` earlier is safe.** I checked it specifically because relocating a call above `renderer.event("prompt_preview", …)` could drop that event on a server-down path — but `preflight.py:183-190` swallows every exception and returns `{}`, so nothing escapes. No event-stream regression.

## Things I liked

- **`_basename` normalizing `\` before `/`.** A pinned bare filename vs. a server-enumerated `SD1.5\name.safetensors` from a Windows host would otherwise miss the basename match and trigger a pointless substitution of the *correct* checkpoint. Both separators are covered by tests (`:254`, `:268`). Easy to miss; good catch.
- **The local/cloud asymmetry on an empty enum.** Hard-erroring locally but failing open on cloud — because Cloud provisions models per-job so an empty cached enum proves nothing — is exactly right, and the reasoning is written down at `preflight.py:147-151` where the next person will find it.
- **`CheckpointResolution` over the ticket's suggested bare `note`.** Your reasoning holds: `None` can't distinguish "pinned present" from "enum positively empty", and those need opposite handling. Right call to deviate.
- Test coverage hits all five branches — substituted, unchanged, empty-enum error, empty-`object_info` fail-open, and user-pinned.

## One correction to the description (not the code)

The **"Judgment calls → `--print-prompt` is unchanged"** bullet says:

> The `prompt_preview` event (emitted before resolution) likewise shows the pre-substitution graph.

That's not what the code does. Resolution runs *before* `renderer.event("prompt_preview", prompt=workflow)` and mutates `workflow` in place, so **`prompt_preview` carries the post-substitution graph** — which your own inline comment states as the intent ("resolves BEFORE the prompt_preview event so the streamed audit trail advertises the graph we actually submit").

The code is doing the better thing; only the description is stale. Worth fixing before squash, since it's a claim about agent-observable stream behavior and a future reader could "fix" the code to match the wrong description.

## Minor nit (non-blocking)

`replacement = enum[0]` is assigned without a type check, while the basename loop just above guards with `isinstance(entry, str)`. A malformed `object_info` with a non-string first element would put a non-string into `ckpt_name`. Not reachable from a well-behaved server, and `_checkpoint_enum` already guards the outer shape — but since the rest of this function is so careful about hostile payloads, an `isinstance(replacement, str)` check (falling through to fail-open otherwise) would match the surrounding standard.
